### ISSUE TITLE
Calender is Blank and Error when trying to call service

### ISSUE THREAD

**Issue by jb123567** [mampfes/hacs_waste_collection_schedule#649]
Hi trying to configure for Wiltshire Council in the UK. I am getting this error when I manually call the service and the calender is blank. Am I missing something?

`websocket_api script: Error executing script. Unexpected error for call_service at pos 1: async_setup.<locals>.fetch_data() takes 0 positional arguments but 1 was given`

`Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 451, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 684, in _async_call_service_step
    await service_task
  File "/usr/src/homeassistant/homeassistant/core.py", line 1755, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1798, in _execute_service
    await self._hass.async_add_executor_job(
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
TypeError: async_setup.<locals>.fetch_data() takes 0 positional arguments but 1 was given`

**Comment by mampfes** [comment]
These are definitely 2 different problems.

1. How do you call `fetch_data`? It doesn't take an argument! But normally you don't have to call this service, esp. if you have problems with missing data.
2. Missing data: Please share your configuration. Did you set `postcode` and `uprn`?

**Comment by jb123567** [comment]
I get the area by simply selecting the fetch data service and calling it in developer tools? I only tried this as the calendar was blank.

My config is as below taken from the readme instructions. I removed my exact details but the post code didn't have spaces and UPRN is correct.

`waste_collection_schedule:
  sources:
    - name: wiltshire_gov_uk
      args:
        postcode: SP****
        uprn: 1009******* `

Thanks,

J

> These are definitely 2 different problems.
> 
> 
> 
> 1. How do you call `fetch_data`? It doesn't take an argument! But normally you don't have to call this service, esp. if you have problems with missing data.
> 
> 2. Missing data: Please share your configuration. Did you set `postcode` and `uprn`?

**Comment by dt215git** [comment]
Using a random SP2 postcode and corresponding uprn it seems to work for me.

```yaml
waste_collection_schedule:
  sources:
    - name: wiltshire_gov_uk
      args:
        postcode: "SP2 ..."
        uprn: "100121......"
      calendar_title: Wilts

sensor:
  - platform: waste_collection_schedule
    name: Wiltshire
    details_format: appointment_types
    date_template: '{{value.date.strftime("%A %d %B %Y")}}'
    add_days_to: true

```
![image](https://user-images.githubusercontent.com/86194065/215837373-d1e54ba8-217f-415b-bac5-a064ce9229fb.png)

![image](https://user-images.githubusercontent.com/86194065/215837786-498013e7-1e1b-483b-8898-ec62ae263137.png)

**Comment by andonevris** [comment]
I'm also getting no data for Wilts.  Using exactly the same config as the example you gave but with SN12 postcode

```
waste_collection_schedule:
  sources:
    - name: wiltshire_gov_uk
      args:
        postcode: "SN12 xxx"
        uprn: "1009xxxxxxx"
      calendar_title: Wilts
```

```
  - platform: waste_collection_schedule
    name: Wiltshire
    details_format: appointment_types
    date_template: '{{value.date.strftime("%A %d %B %Y")}}'
    add_days_to: true
```

![image](https://user-images.githubusercontent.com/1779529/215970976-71dba2a2-edc2-48ec-904c-a54494016a65.png)

**Comment by jb123567** [comment]
> 

Using the same config as you but changing the UPRN and post code still gets me the error above and no data.

**Comment by jb123567** [comment]
> I'm also getting no data for Wilts. Using exactly the same config as the example you gave but with SN12 postcode
> 
> ```
> waste_collection_schedule:
>   sources:
>     - name: wiltshire_gov_uk
>       args:
>         postcode: "SN12 xxx"
>         uprn: "1009xxxxxxx"
>       calendar_title: Wilts
> ```
> 
> ```
>   - platform: waste_collection_schedule
>     name: Wiltshire
>     details_format: appointment_types
>     date_template: '{{value.date.strftime("%A %d %B %Y")}}'
>     add_days_to: true
> ```
> 
> ![image](https://user-images.githubusercontent.com/1779529/215970976-71dba2a2-edc2-48ec-904c-a54494016a65.png)

I've got it working on a different but local address. By any chance is your address relatively new? Mine is 2016 and doesnt work. An older address, say 1990's works. Must be a Wilts council issue.

**Comment by andonevris** [comment]
> > I'm also getting no data for Wilts. Using exactly the same config as the example you gave but with SN12 postcode
> > ```
> > waste_collection_schedule:
> >   sources:
> >     - name: wiltshire_gov_uk
> >       args:
> >         postcode: "SN12 xxx"
> >         uprn: "1009xxxxxxx"
> >       calendar_title: Wilts
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> >   - platform: waste_collection_schedule
> >     name: Wiltshire
> >     details_format: appointment_types
> >     date_template: '{{value.date.strftime("%A %d %B %Y")}}'
> >     add_days_to: true
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ![image](https://user-images.githubusercontent.com/1779529/215970976-71dba2a2-edc2-48ec-904c-a54494016a65.png)
> 
> I've got it working on a different but local address. By any chance is your address relatively new? Mine is 2016 and doesnt work. An older address, say 1990's works. Must be a Wilts council issue.

Actually yes my house is 3 years old, that might be it. 

I can manually search for my address at the wilts council website and it finds my address and displays the calendar fine.

**Comment by dt215git** [comment]
Some random SN12 postcodes also seem to work for me.

Pretty much every house in the UK should have a uprn, and the underlying query on the Wilts website is using it to pull back the calendar it displays. If you find your uprn from [findmyaddress](https://www.findmyaddress.co.uk/search), then that should also be what their website is using.

The test case for Wilt uses:
- BA149QP
- 100121085972

Can you get that working on your HA instance? If that starts displaying, then we've ruled out config issues.

@jb123567: Do you get a collection calendar on the Wilts web site?
@andonevris: If you can see the calendar on the website then perhaps there's  something odd going on with the config.

**Comment by andonevris** [comment]
> @andonevris: If you can see the calendar on the website then perhaps there's something odd going on with the config.

The test case works fine (BTW I could only get it to work with no quote marks and no space in the postcode)

When I plug my property data back in the calendar is blank.

**Comment by dt215git** [comment]
@andonevris: the need for `"` is a bit hit and miss - in the underlying code sometimes numbers need to be numbers and sometimes they need to be strings, not all scripts explicitly convert them to the format they need. The type of HA you're running with might also make a difference. At least you've figured out what works for you.

Are there any other postcodes from your surrounding new-build area that fail in HA but work on the Wilts website? I can try those and see if they work/fail on my set up.

Other gotcha's are:

- Not clearing/refreshing your browser cache after making changes
- Having a disabled calendar/sensor entity without realising it
- Not spotting error messages in the log files

**Comment by andonevris** [comment]
> Are there any other postcodes from your surrounding new-build area that fail in HA but work on the Wilts website? I can try those and see if they work/fail on my set up.

There is a new build area near me that's around 6 years old, I just checked a random house from there and having the same blank calendar.

        postcode: SN128FF
        uprn: 10093279004

**Comment by dt215git** [comment]
That uprn returns a blank calendar on the Wilts web site, so not surprising nothing shows up in HA.
![image](https://user-images.githubusercontent.com/86194065/216100730-6d95e030-e991-4b95-8ca1-f7dd92badffa.png)

Changing the uprn to the neighbouring house gives me this:
![image](https://user-images.githubusercontent.com/86194065/216102197-f94c9884-19c6-438f-8ed1-e16cde63f5b8.png)
![image](https://user-images.githubusercontent.com/86194065/216102392-5ddbae7c-d167-48b8-a6fc-1cb65fde616d.png)

However, it looks like the uprn has to be 12 digits long - the website is padding the uprn with a leading `0`. Maybe all the ones I tried previously were already 12 digits. If you do that to your own uprn what happens?

**Comment by andonevris** [comment]
Yes success.  Adding a leading zero to my 11 digit UPRN fixed it.

Thanks for your help, good catch!

**Comment by dt215git** [comment]
Yay! I'll create a PR for the source so it automatically pads uprn to 12 digits.

**Comment by jb123567** [comment]
> Yay! I'll create a PR for the source so it automatically pads uprn to 12 digits.

Adding the 0 solved it for me too!!

**Comment by mampfes** [comment]
fixed by #656 

Thanks everyone, great team work!!!
